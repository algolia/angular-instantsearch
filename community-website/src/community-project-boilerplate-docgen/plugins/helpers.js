const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
