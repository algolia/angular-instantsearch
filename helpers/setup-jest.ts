import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
