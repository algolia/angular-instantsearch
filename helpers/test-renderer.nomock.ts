import { Component, ViewChild } from '@angular/core';
import { TestBed } from '@angular/core/testing';

import { NgAisInstantSearchModule } from '../src/instantsearch/instantsearch.module';

// same as test-renderer, but without mocking InstantSearch
// In the future, we want to migrate all tests to this

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations,
  methods = {},
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
  methods?: {};
}) {
  return function(state?: {}, isFirstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined,
        methods,
      },
      isFirstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state,
    methods,
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
    methods?: {};
  },
  isFirstRender = false
) {
  @Component({
    template: `
      <ais-instantsearch [config]="config">
        ${template}
      </ais-instantsearch>
    `,
  })
  class TestContainer {
    config = {
      searchClient: {
        search(requests) {
          return Promise.resolve({
            results: requests.map(() => ({ hits: [] })),
          });
        },
      },
      indexName: 'instant_search',
    };
    @ViewChild(TestedWidget) testedWidget;
    constructor() {
      Object.keys(methods).forEach(methodName => {
        this[methodName] = methods[methodName];
      });
    }
  }

  TestBed.configureCompiler({
    preserveWhitespaces: false,
  } as any).configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || []),
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])],
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, isFirstRender);
    fixture.detectChanges();
  }

  return fixture;
}
