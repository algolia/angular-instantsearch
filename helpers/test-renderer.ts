import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
