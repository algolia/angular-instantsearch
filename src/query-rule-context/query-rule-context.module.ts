import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgAisQueryRuleContext } from './query-rule-context';

@NgModule({
  declarations: [NgAisQueryRuleContext],
  entryComponents: [NgAisQueryRuleContext],
  exports: [NgAisQueryRuleContext],
  imports: [CommonModule],
})
export class NgAisQueryRuleContextModule {}
