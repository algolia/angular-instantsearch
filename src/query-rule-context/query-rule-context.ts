import { Component, Input, Inject, forwardRef } from '@angular/core';

import { connectQueryRules } from 'instantsearch.js/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgAisInstantSearch } from '../instantsearch/instantsearch';

type FacetValue = string | number | boolean;

@Component({
  selector: 'ais-query-rule-context',
  template: '',
})
export class NgAisQueryRuleContext extends BaseWidget {
  @Input()
  public trackedFilters: {
    [facetName: string]: (facetValues: FacetValue[]) => FacetValue[];
  };
  @Input() public transformRuleContexts?: (items: string[]) => string[];

  constructor(
    @Inject(forwardRef(() => NgAisInstantSearch))
    public instantSearchParent: any
  ) {
    super('QueryRuleContext');
  }

  public ngOnInit() {
    this.createWidget(connectQueryRules, {
      trackedFilters: this.trackedFilters,
      transformRuleContexts: this.transformRuleContexts,
    });

    super.ngOnInit();
  }
}
