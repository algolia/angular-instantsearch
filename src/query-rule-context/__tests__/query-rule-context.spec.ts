import { createRenderer } from '../../../helpers/test-renderer';
import { NgAisQueryRuleContext } from '../query-rule-context';

describe('QueryRuleContext', () => {
  let render;
  let createWidget;

  beforeEach(() => {
    jest.clearAllMocks();

    createWidget = jest.spyOn(NgAisQueryRuleContext.prototype, 'createWidget');

    render = (template: string, state?: object) =>
      createRenderer({
        defaultState: state,
        template,
        TestedWidget: NgAisQueryRuleContext,
      })({});
  });

  it('renders nothing without state', () => {
    const template = '<ais-query-rule-context></ais-query-rule-context>';
    const state = undefined;

    const fixture = render(template, state);

    expect(fixture).toMatchSnapshot();
  });

  it('renders nothing with state', () => {
    const template = '<ais-query-rule-context></ais-query-rule-context>';
    const state = {
      items: ['Luke', 'I am', 'your father'],
    };

    const fixture = render(template, state);

    expect(fixture).toMatchSnapshot();
  });

  it('passes trackedFilters', () => {
    const trackedFilters = 'trackedFilters';
    const template = `<ais-query-rule-context
      [trackedFilters]="'${trackedFilters}'"
    >
    </ais-query-rule-context>`;
    const state = {};

    render(template, state);

    expect(createWidget.mock.calls[0][1].trackedFilters).toEqual(
      trackedFilters
    );
  });

  it('passes transformRuleContexts', () => {
    const transformRuleContexts = 'transformRuleContexts';
    const template = `<ais-query-rule-context
      [transformRuleContexts]="'${transformRuleContexts}'"
    >
    </ais-query-rule-context>`;
    const state = {};

    render(template, state);

    expect(createWidget.mock.calls[0][1].transformRuleContexts).toEqual(
      transformRuleContexts
    );
  });
});
