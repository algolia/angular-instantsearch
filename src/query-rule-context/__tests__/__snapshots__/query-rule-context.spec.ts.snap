// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryRuleContext renders nothing with state 1`] = `
<ng-component
  testedWidget={[Function NgAisQueryRuleContext]}
>
  <ais-instantsearch>
    <ais-query-rule-context />
  </ais-instantsearch>
</ng-component>
`;

exports[`QueryRuleContext renders nothing without state 1`] = `
<ng-component
  testedWidget={[Function NgAisQueryRuleContext]}
>
  <ais-instantsearch>
    <ais-query-rule-context />
  </ais-instantsearch>
</ng-component>
`;
